### PR TITLE
Feature/9 zero3 and datasets

### LINKED ISSUE #9: Train with deepspeed ZeRO3 and huggingface datasets to load data
## Describe a requested feature
- training code for using zero3
- load dataset with datasets library for faster encoding and caching

### PR DESCRIPTION
**Pull request by bzantium**
- resolved #9

### CODE REVIEW

File: ahn/training/train_zero3.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+                    data={
+                        "train_loss": loss.item(),
+                        "train_ppl": math.exp(loss.item()),
+                    },
+                    step=config["training"]["current_step"],
+                )
+
+            engine.backward(loss)
+            engine.step()
+            torch.cuda.empty_cache()
```
Comment (hyunwoongko) [inline ahn/training/train_zero3.py]: should we empty gpu cache?

File: ahn/training/_decoder.py [hunk 2]
```diff
@@ -49,34 +51,140 @@ def tokenize(config, tokenizer, data, data_type):
...
+            padding=False,
+            truncation=False,
+            max_length=config["model_and_tokenizer"]["max_length"],
+            return_attention_mask=False,
+            return_token_type_ids=False,
+        )
+        return result
+
+    with accelerator.main_process_first():
+        dataset = dataset.map(
```
Comment (hyunwoongko) [inline ahn/training/_decoder.py:83]: what's the advantage of this?

Comment (bzantium) [reply ahn/training/_decoder.py:83]: first of all, we don't need to retokenize because it uses caching after map at first. Also by using batch tokenizing, it would make faster.

File: ahn/training/train_zero3.py (hunk 1 above)
Comment (bzantium) [reply ahn/training/train_zero3.py]: I added it becuase the following waring:
```pytorch allocator cache flushes since last step. this happens when there is high memory pressure and is detrimental to performance. if this is happening frequently consider adjusting settings to reduce memory consumption. If you are unable to make the cache flushes go away consider adding get_accelerator().empty_cache() calls in your training loop to ensure that all ranks flush their caches at the same time```
with experiment, I find that there's not much effect on training time but still it does not resolve the warninig above.

File: ahn/training/_decoder.py [hunk 3]
```diff
@@ -1,11 +1,13 @@
-import json
+import jsonlines
```
Comment (jason9693) [inline ahn/training/_decoder.py:1]: Is it a lot of differ with using json lib?

File: ahn/training/_decoder.py [hunk 4]
```diff
@@ -124,7 +232,7 @@ def collate_fn(batch):
     return DataLoader(
         dataset,
         batch_size=config["deepspeed"]["train_micro_batch_size_per_gpu"],
-        num_workers=os.cpu_count() // dist.get_world_size(),
+        num_workers=4,
```
Comment (jason9693) [inline ahn/training/_decoder.py:236]: Why did you using static value(4) instead of follow by world size?

Comment (bzantium) [reply ahn/training/_decoder.py:236]: Because of following warning:
```UserWarning: This DataLoader will create 12 worker processes in total. Our suggested max number of worker in current system is 4, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.```

File: ahn/training/_decoder.py (hunk 3 above)
Comment (bzantium) [reply ahn/training/_decoder.py:1]: Not so much different I think... just for better code interface
